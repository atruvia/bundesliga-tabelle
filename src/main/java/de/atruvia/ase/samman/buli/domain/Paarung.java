package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.GEPLANT;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.LAUFEND;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;

import java.util.function.Function;

import org.jmolecules.ddd.annotation.ValueObject;

import lombok.Builder;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Builder(toBuilder = true)
@Accessors(fluent = true)
@ValueObject
public class Paarung {

	@Value
	public class PaarungView {

		ViewDirection direction;
		Function<Paarung, Entry> self;
		Function<Paarung, Entry> gegner;
		@Getter(lazy = true)
		Ergebnis ergebnis = calcErgebnis();

		public Entry self() {
			return self.apply(paarung());
		}

		public Entry gegner() {
			return gegner.apply(paarung());
		}

		public int tore() {
			return self().tore();
		}

		public int gegentore() {
			return gegner().tore();
		}

		public boolean isGeplant() {
			return paarung().isGeplant();
		}

		public boolean isLaufend() {
			return paarung().isLaufend();
		}

		private Ergebnis calcErgebnis() {
			var tore = tore();
			var gegentore = gegentore();
			if (tore == gegentore) {
				return UNENTSCHIEDEN;
			} else if (tore > gegentore) {
				return SIEG;
			} else {
				return NIEDERLAGE;
			}
		}

		public ErgebnisTyp ergebnisTyp() {
			return paarung().ergebnisTyp;
		}

		private Paarung paarung() {
			return Paarung.this;
		}

	}

	public enum ViewDirection {
		HEIM, AUSWAERTS
	}

	@RequiredArgsConstructor
	@Getter
	@Accessors(fluent = true)
	public enum Ergebnis {

		SIEG, UNENTSCHIEDEN, NIEDERLAGE;

		public int punkte() {
			return switch (this) {
			case SIEG -> 3;
			case UNENTSCHIEDEN -> 1;
			case NIEDERLAGE -> 0;
			};
		}

	}

	public enum ErgebnisTyp {
		GEPLANT, LAUFEND, BEENDET
	}

	@Value
	@Builder(toBuilder = true)
	public static class Entry {
		Team team;
		@With
		int tore;

		public Entry(Team team, int tore) {
			if (tore < 0) {
				throw new IllegalArgumentException("Tore darf nicht negativ sein, ist aber " + tore);
			}
			this.team = team;
			this.tore = tore;
		}

	}

	@Builder.Default
	ErgebnisTyp ergebnisTyp = GEPLANT;

	Entry heim, gast;

	public boolean isGeplant() {
		return ergebnisTypIs(GEPLANT);
	}

	public boolean isLaufend() {
		return ergebnisTypIs(LAUFEND);
	}

	private boolean ergebnisTypIs(ErgebnisTyp ergebnisTyp) {
		return this.ergebnisTyp == ergebnisTyp;
	}

	public Paarung withErgebnis(int toreHeim, int toreGast) {
		return toBuilder().endergebnis(toreHeim, toreGast).build();
	}

	public ViewDirection viewDirection() {
		return HEIM;
	}

	public static class PaarungBuilder {

		public static PaarungBuilder paarung(String heim, String gast) {
			return paarung(team(heim), team(gast));
		}

		private static Team team(String name) {
			return Team.builder().name(name).build();
		}

		public static PaarungBuilder paarung(Team heim, Team gast) {
			return Paarung.builder().heim(entry(heim)).gast(entry(gast));
		}

		public PaarungBuilder endergebnis(int toreHeim, int toreGast) {
			return ergebnis(BEENDET, toreHeim, toreGast);
		}

		public PaarungBuilder zwischenergebnis(int toreHeim, int toreGast) {
			return ergebnis(LAUFEND, toreHeim, toreGast);
		}

		PaarungBuilder ergebnis(ErgebnisTyp ergebnisTyp, int toreHeim, int toreGast) {
			return ergebnisTyp(ergebnisTyp).goals(toreHeim, toreGast);
		}

		private static Entry entry(Team team) {
			return Entry.builder().team(team).build();
		}

		public PaarungBuilder goals(int toreHeim, int toreGast) {
			return heim(heim.withTore(toreHeim)).gast(gast.withTore(toreGast));
		}

	}

	public PaarungView viewForTeam(ViewDirection viewDirection) {
		return switch (viewDirection) {
		case HEIM -> new PaarungView(HEIM, Paarung::heim, Paarung::gast);
		case AUSWAERTS -> new PaarungView(AUSWAERTS, Paarung::gast, Paarung::heim);
		};
	}

}