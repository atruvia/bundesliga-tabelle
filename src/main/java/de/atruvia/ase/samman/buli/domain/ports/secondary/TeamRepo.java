package de.atruvia.ase.samman.buli.domain.ports.secondary;

import java.net.URI;

public interface TeamRepo {

	URI getTeams(String league, String season, String team) throws Exception;

	URI getTeams(String league, String season) throws Exception;

}
