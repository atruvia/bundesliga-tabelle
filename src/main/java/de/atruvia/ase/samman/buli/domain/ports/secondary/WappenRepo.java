package de.atruvia.ase.samman.buli.domain.ports.secondary;

import java.net.URI;

public interface WappenRepo {

	URI getWappen(String league, String season, String team) throws Exception;

}
