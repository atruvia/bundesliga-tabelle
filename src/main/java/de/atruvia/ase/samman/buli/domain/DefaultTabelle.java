package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Stream.concat;
import static lombok.AccessLevel.PRIVATE;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

import org.jmolecules.ddd.annotation.ValueObject;

import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.experimental.Accessors;

@RequiredArgsConstructor
@ValueObject
public class DefaultTabelle implements Tabelle {

	@Value
	@Accessors(fluent = true)
	private static class OrdnungsElement implements Comparable<OrdnungsElement> {

		private static final Comparator<OrdnungsElement> comparator = comparing(value(TabellenPlatz::punkte)) //
				.thenComparing(value(TabellenPlatz::torDifferenz)) //
				.thenComparing(value(TabellenPlatz::gesamtTore)) //
				.thenComparing(direkterVergleichGesamt()) //
				.thenComparing(direkterVergleichAuswaertsTore()) //
				.thenComparing(value(TabellenPlatz::auswaertsTore)) //
				.reversed();

		private static Comparator<OrdnungsElement> direkterVergleichGesamt() {
			return compareWithSwapped((that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team())));
		}

		private static Comparator<OrdnungsElement> direkterVergleichAuswaertsTore() {
			return compareWithSwapped(
					(that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team()).and(isAuswaerts())));
		}

		private static <T, R extends Comparable<R>> Comparator<T> compareWithSwapped(BiFunction<T, T, R> biFunction) {
			return (o1, o2) -> biFunction.apply(o1, o2).compareTo(biFunction.apply(o2, o1));
		}

		private Team team() {
			return tabellenPlatz.team();
		}

		private static Predicate<PaarungView> gegnerIs(Team gegner) {
			return e -> Objects.equals(e.gegner().team().id(), gegner.id());
		}

		private static Predicate<PaarungView> isAuswaerts() {
			return e -> Objects.equals(e.direction(), AUSWAERTS);
		}

		private static int whereToreIs(TabellenPlatz tabellenPlatz, Predicate<PaarungView> filter) {
			return tabellenPlatz.paarungenStream().filter(filter).map(PaarungView::tore).mapToInt(Integer::valueOf)
					.sum();
		}

		@Getter(value = PRIVATE)
		TabellenPlatz tabellenPlatz;

		@Override
		public int hashCode() {
			return 0;
		}

		private static <T> Function<OrdnungsElement, T> value(Function<TabellenPlatz, T> function) {
			return t -> function.apply(t.tabellenPlatz);
		}

		@Override
		public boolean equals(Object o) {
			return comparator.compare(this, (OrdnungsElement) o) == 0;
		}

		@Override
		public int compareTo(OrdnungsElement other) {
			return comparator.thenComparing(e -> e.team().name()).compare(this, other);
		}

	}

	private final Collection<TabellenPlatz> entries;

	public DefaultTabelle() {
		this(emptyList());
	}

	@Override
	public Tabelle add(Paarung paarung) {
		var existing = entries().stream();
		var toAdd = Stream.of(newEntry(paarung.viewForTeam(HEIM)), newEntry(paarung.viewForTeam(AUSWAERTS)));
		var newEntries = concat(existing, toAdd) //
				.collect(toMap(t -> t.team().id(), identity(), TabellenPlatz::mergeWith)) //
				.values();
		return new DefaultTabelle(newEntries);
	}

	private static TabellenPlatz newEntry(PaarungView paarung) {
		var builder = TabellenPlatz.builder().team(paarung.self().team());
		return (paarung.isGeplant() ? builder : builder.paarung(paarung)).build();
	}

	public List<TabellenPlatz> entries() {
		// TODO make it side-effect-free
		AtomicInteger counter = new AtomicInteger(1);
		Map<OrdnungsElement, List<TabellenPlatz>> platzGruppen = entries.stream()
				.collect(groupingBy(OrdnungsElement::new));
		return new TreeMap<>(platzGruppen).values().stream() //
				.flatMap(t -> makeGroup(counter, t)) //
				.toList();
	}

	private static Stream<TabellenPlatz> makeGroup(AtomicInteger counter, List<TabellenPlatz> tabellenPlaetze) {
		Rank rank = new Rank(counter.getAndAdd(tabellenPlaetze.size()));
		return tabellenPlaetze.stream() //
				.sorted(comparing(OrdnungsElement::new)) //
				.map(tp -> tp.withRank(rank));
	}

}
