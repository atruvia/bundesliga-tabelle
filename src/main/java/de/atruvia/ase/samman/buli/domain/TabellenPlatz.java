package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static de.atruvia.ase.samman.buli.util.Merger.lastNonNull;
import static de.atruvia.ase.samman.buli.util.Merger.merge;
import static java.util.Arrays.asList;
import static java.util.stream.IntStream.range;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BinaryOperator;
import java.util.stream.Stream;

import de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis;
import de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp;
import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection;
import de.atruvia.ase.samman.buli.util.Merger.Mergeable;
import lombok.Builder;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Accessors(fluent = true)
@Builder
public class TabellenPlatz implements Mergeable<TabellenPlatz> {

	@Value
	private static class ErgebnisEntry {
		Ergebnis ergebnis;
		ErgebnisTyp ergebnisTyp;
	}

	private static final BinaryOperator<Integer> adder = (i1, i2) -> i1 + i2;

	URI wappen;
	@With
	int platz;
	@With
	String team;
	int spiele;
	List<ErgebnisEntry> ergebnisse;
	int punkte;
	Map<ViewDirection, Integer> tore;
	Map<ViewDirection, Integer> gegentore;
	PaarungView laufendesSpiel;

	public List<Ergebnis> ergebnisse() {
		return collectToList(ergebnisseStream());
	}

	public List<Ergebnis> ergebnisse(ErgebnisTyp... ergebnisTyp) {
		return collectToList(ergebnisseStream().filter(e -> entryErgebnisIsTypeOf(e, ergebnisTyp)));
	}

	private Stream<ErgebnisEntry> ergebnisseStream() {
		return ergebnisse.stream();
	}

	private static List<Ergebnis> collectToList(Stream<ErgebnisEntry> filter) {
		return filter.map(ErgebnisEntry::ergebnis).toList();
	}

	private static boolean entryErgebnisIsTypeOf(ErgebnisEntry e, ErgebnisTyp... ergebnisTyp) {
		return asList(ergebnisTyp).contains(e.ergebnisTyp());
	}

	public int gesamtTore() {
		return heimtore() + auswaertsTore();
	}

	public int gesamtGegentore() {
		return heimGegentore() + auswaertsGegentore();
	}

	public int heimtore() {
		return tore().getOrDefault(HEIM, 0);
	}

	public int auswaertsTore() {
		return tore().getOrDefault(AUSWAERTS, 0);
	}

	public int heimGegentore() {
		return gegentore().getOrDefault(HEIM, 0);
	}

	public int auswaertsGegentore() {
		return gegentore().getOrDefault(AUSWAERTS, 0);
	}

	public int torDifferenz() {
		return gesamtTore() - gesamtGegentore();
	}

	public TabellenPlatzBuilder toBuilder() {
		TabellenPlatzBuilder builder = new TabellenPlatzBuilder();
		builder.wappen = wappen;
		builder.platz = platz;
		builder.team = team;
		builder.spiele = spiele;
		builder.ergebnisse = new ArrayList<>(ergebnisse);
		builder.punkte = punkte;
		builder.tore = new HashMap<>(tore);
		builder.gegentore = new HashMap<>(gegentore);
		builder.laufendesSpiel = laufendesSpiel;
		return builder;
	}

	public static class TabellenPlatzBuilder {

		public TabellenPlatzBuilder() {
			ergebnisse = new ArrayList<>();
			tore = new HashMap<>();
			gegentore = new HashMap<>();
		}

		public TabellenPlatzBuilder ergebnis(Ergebnis ergebnis, ErgebnisTyp ergebnisTyp) {
			ergebnisse.add(new ErgebnisEntry(ergebnis, ergebnisTyp));
			return this;
		}

		public TabellenPlatzBuilder withGegentore(ViewDirection direction, int anzahl) {
			gegentore.put(direction, anzahl);
			return this;
		}

		public TabellenPlatzBuilder withTore(ViewDirection viewDirection, int anzahl) {
			tore.put(viewDirection, anzahl);
			return this;
		}

	}

	@Override
	public TabellenPlatz mergeWith(TabellenPlatz other) {
		return builder() //
				.team(lastNonNull(team, other.team)) //
				.ergebnisse(merge(ergebnisse, other.ergebnisse)) //
				.spiele(merge(spiele, other.spiele)) //
				.punkte(merge(punkte, other.punkte)) //
				.tore(merge(adder, tore, other.tore)) //
				.gegentore(merge(adder, gegentore, other.gegentore)) //
				.wappen(lastNonNull(wappen, other.wappen)) //
				.laufendesSpiel(lastNonNull(laufendesSpiel, other.laufendesSpiel)) //
				.build();
	}

	public int siege() {
		return countAnzahl(SIEG);
	}

	public int unentschieden() {
		return countAnzahl(UNENTSCHIEDEN);
	}

	public int niederlagen() {
		return countAnzahl(NIEDERLAGE);
	}

	private int countAnzahl(Ergebnis type) {
		return (int) ergebnisseStream().map(ErgebnisEntry::ergebnis).filter(type::equals).count();
	}

	public Ergebnis[] tendenz() {
		return last(5, ergebnisse(BEENDET)).toArray(Ergebnis[]::new);
	}

	private static <T> Stream<T> last(int count, List<T> elements) {
		int size = elements.size();
		return range(0, count).mapToObj(i -> {
			int index = size - i - 1;
			return index >= 0 ? elements.get(index) : null;
		});
	}

}