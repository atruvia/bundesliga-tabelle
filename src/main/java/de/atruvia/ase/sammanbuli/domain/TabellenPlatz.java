package de.atruvia.ase.sammanbuli.domain;

import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static java.util.Collections.emptyMap;
import static java.util.stream.Collectors.toMap;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Stream;

import de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis;
import lombok.Builder;
import lombok.Value;
import lombok.With;

@Value
@Builder
public class TabellenPlatz {

	static TabellenPlatz NULL = new TabellenPlatz(0, "", 0, emptyMap(), 0, 0, 0, 0, 0);

	@With
	int platz;
	@With
	String team;
	@Builder.Default
	int spiele = 1;
	Map<Ergebnis, Integer> ergebnisse;
	int punkte;
	int toreHeim;
	int toreAuswaerts;
	int gegentoreHeim;
	int gegentoreAuswaerts;

	public int getTore() {
		return toreHeim + toreAuswaerts;
	}

	public int getGegentore() {
		return gegentoreHeim + gegentoreAuswaerts;
	}

	static class TabellenPlatzBuilder {

		TabellenPlatzBuilder() {
			ergebnisse = new HashMap<>();
		}

		public TabellenPlatz.TabellenPlatzBuilder ergebnis(Ergebnis ergebnis) {
			ergebnisse.merge(ergebnis, 1, (a, b) -> a + b);
			return this;
		}

	}

	public int getTorDifferenz() {
		return getTore() - getGegentore();
	}

	public TabellenPlatz merge(TabellenPlatz other) {
		return TabellenPlatz.builder() //
				.ergebnisse(merge(this.ergebnisse, other.ergebnisse)) //
				.spiele(this.spiele + other.spiele) //
				.punkte(this.punkte + other.punkte) //
				.toreHeim(this.toreHeim + other.toreHeim) //
				.toreAuswaerts(this.toreAuswaerts + other.toreAuswaerts) //
				.gegentoreHeim(this.gegentoreHeim + other.gegentoreHeim) //
				.gegentoreAuswaerts(this.gegentoreAuswaerts + other.gegentoreAuswaerts) //
				.build();
	}

	private static Map<Ergebnis, Integer> merge(Map<Ergebnis, Integer> map1, Map<Ergebnis, Integer> map2) {
		return Stream.of(map1, map2).map(Map::entrySet).flatMap(Collection::stream)
				.collect(toMap(Entry::getKey, Entry::getValue, Integer::sum));
	}

	public int getGewonnen() {
		return ergebnis(SIEG);
	}

	public int getUnentschieden() {
		return ergebnis(UNENTSCHIEDEN);
	}

	public int getVerloren() {
		return ergebnis(NIEDERLAGE);
	}

	private int ergebnis(Ergebnis type) {
		return ergebnisse.getOrDefault(type, 0);
	}

}