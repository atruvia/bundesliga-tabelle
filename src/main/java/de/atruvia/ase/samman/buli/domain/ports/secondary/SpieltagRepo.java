package de.atruvia.ase.samman.buli.domain.ports.secondary;

import java.util.List;

import org.jmolecules.architecture.hexagonal.SecondaryPort;

import de.atruvia.ase.samman.buli.domain.Paarung;

@SecondaryPort
public interface SpieltagRepo {

	List<Paarung> lade(String league, String season);

}
