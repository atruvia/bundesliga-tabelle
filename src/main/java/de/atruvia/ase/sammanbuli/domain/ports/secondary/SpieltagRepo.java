package de.atruvia.ase.sammanbuli.domain.ports.secondary;

import java.util.List;

import de.atruvia.ase.samman.buli.domain.Paarung;

public interface SpieltagRepo {

	List<Paarung> lade(String league, String season) throws Exception;

}
