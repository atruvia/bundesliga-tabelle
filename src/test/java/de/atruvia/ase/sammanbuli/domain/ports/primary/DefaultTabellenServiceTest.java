package de.atruvia.ase.sammanbuli.domain.ports.primary;

import static de.atruvia.ase.sammanbuli.infra.adapters.secondary.OpenLigaDbSpieltagRepoMother.readFromLocalFilesystemRepo;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.fail;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.junit.jupiter.api.Test;

import de.atruvia.ase.sammanbuli.domain.TabellenPlatz;
import de.atruvia.ase.sammanbuli.domain.ports.secondary.SpieltagRepo;

class DefaultTabellenServiceTest {

	@Test
	void test() {
		TabellenService sut = new DefaultTabellenService(readFromLocalFilesystemRepo());
		List<TabellenPlatz> erstelleTabelle = sut.erstelleTabelle("bl1", "2022");

		String s = erstelleTabelle.stream().map(f -> print(f)).collect(Collectors.joining("\n"));

		assertThat(s).isEqualTo("");
	}

	private String print(TabellenPlatz tabellenPlatz) {
		tabellenPlatz.
		// TODO Auto-generated method stub
		return null;
	}

}
