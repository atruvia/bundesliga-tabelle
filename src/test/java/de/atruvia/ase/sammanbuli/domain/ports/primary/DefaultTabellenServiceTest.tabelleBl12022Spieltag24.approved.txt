FC Bayern München         |  24 |  15 |   7 |   2 |  71 |  25 |  46 |  52
Borussia Dortmund         |  24 |  16 |   2 |   6 |  49 |  30 |  19 |  50
RB Leipzig                |  24 |  13 |   6 |   5 |  49 |  29 |  20 |  45
1. FC Union Berlin        |  24 |  13 |   6 |   5 |  36 |  28 |   8 |  45
SC Freiburg               |  24 |  13 |   6 |   5 |  37 |  33 |   4 |  45
Eintracht Frankfurt       |  24 |  11 |   7 |   6 |  46 |  34 |  12 |  40
1. FSV Mainz 05           |  24 |  10 |   6 |   8 |  39 |  35 |   4 |  36
VfL Wolfsburg             |  24 |   9 |   8 |   7 |  43 |  32 |  11 |  35
Bayer Leverkusen          |  24 |  10 |   4 |  10 |  43 |  39 |   4 |  34
Borussia Mönchengladbach  |  24 |   8 |   6 |  10 |  38 |  42 |  -4 |  30
Werder Bremen             |  24 |   9 |   3 |  12 |  37 |  46 |  -9 |  30
1. FC Köln                |  24 |   6 |   9 |   9 |  32 |  38 |  -6 |  27
FC Augsburg               |  24 |   8 |   3 |  13 |  31 |  44 | -13 |  27
VfL Bochum                |  24 |   7 |   1 |  16 |  26 |  56 | -30 |  22
Hertha BSC                |  24 |   5 |   6 |  13 |  29 |  45 | -16 |  21
VfB Stuttgart             |  24 |   4 |   8 |  12 |  29 |  41 | -12 |  20
FC Schalke 04             |  24 |   4 |   8 |  12 |  20 |  44 | -24 |  20
TSG 1899 Hoffenheim       |  24 |   5 |   4 |  15 |  30 |  44 | -14 |  19
